### ISSUE TITLE
Tensorboard selective purge (wacky graphs)

### ISSUE THREAD

**Issue by PetrochukM** [lanpa/tensorboard-pytorch#185]
Tensorboard purges selectively for one graph but not the other, minimal code example below...

**Graphs**
![image](https://user-images.githubusercontent.com/7424737/42385200-f32525ca-80f0-11e8-9b71-e0658f3113ec.png)

**Output**
```
OS-XImage-9:Tacotron-2 michaelp$ tensorboard --logdir=welp
/Users/michaelp/.pyenv/versions/3.6.5/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
2018-07-06 07:45:18.693751: I tensorflow/core/platform/cpu_feature_guard.cc:140] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
W0706 07:45:18.702591 Reloader tf_logging.py:121] Detected out of order event.step likely caused by a TensorFlow restart. Purging 2 expired tensor events from Tensorboard display between the previous step: 2 (timestamp: 1530888314.12851) and currentstep: 1 (timestamp: 1530888315.1338508).
TensorBoard 1.8.0 at http://OS-XImage-9:6006 (Press CTRL+C to quit)
```

**Code**
```python
from time import sleep

import shutil
import os

from tensorboardX import SummaryWriter
from tensorboardX.src.event_pb2 import SessionLog

if os.path.isdir('welp'):
    shutil.rmtree('welp')

val = 3

tensorboard = SummaryWriter(log_dir='welp/train')
tensorboard.file_writer.add_session_log(
    session_log=SessionLog(status=SessionLog.START), global_step=0)
tensorboard.add_scalar('path/one/', 0, 0)
tensorboard.add_scalar('path/two/', 0, 0)
tensorboard.add_scalar('path/one/', val, 1)
tensorboard.add_scalar('path/two/', val, 1)
tensorboard.add_scalar('path/one/', 0, 2)
tensorboard.add_scalar('path/two/', 0, 2)
tensorboard.close()

sleep(1.0)

tensorboard = SummaryWriter(log_dir='welp/train')
tensorboard.file_writer.add_session_log(
    session_log=SessionLog(status=SessionLog.START), global_step=1)
tensorboard.add_scalar('path/one/', 0, 1)
tensorboard.add_scalar('path/two/', 0, 1)
tensorboard.close()
```

**Inspect**
```bash
OS-XImage-9:Tacotron-2 michaelp$ tensorboard --inspect --logdir=welp
/Users/michaelp/.pyenv/versions/3.6.5/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
======================================================================
Processing event files... (this can take a few minutes)
======================================================================

Found event files in:
welp/train

These tags are in welp/train:
audio -
histograms -
images -
scalars
   path/one/
   path/two/
tensor -
======================================================================

Event statistics for welp/train:
audio -
graph -
histograms -
images -
scalars
   first_step           0
   last_step            1
   max_step             2
   min_step             0
   num_steps            3
   outoforder_steps     [(2, 1)]
sessionlog:checkpoint -
sessionlog:start
   outoforder_steps     []
   steps                [0, 1]
sessionlog:stop -
tensor -
======================================================================
```

**System**
 - TensorBoard 1.8.0
 - macOS High Sierra 
 - 3.5
 
Reposted here: https://github.com/tensorflow/tensorboard/issues/1272

**Comment by PetrochukM** [comment]
Figured it out:
```
from time import sleep

import shutil
import os

from tensorboardX import SummaryWriter
from tensorboardX.src.event_pb2 import SessionLog
from tensorboardX.src.event_pb2 import Event

if os.path.isdir('welp'):
    shutil.rmtree('welp')

val = 3

tensorboard = SummaryWriter(log_dir='welp/train')
tensorboard.file_writer.event_writer.add_event(Event(step=0, file_version='brain.Event:2'))
tensorboard.file_writer.event_writer.add_event(
    Event(session_log=SessionLog(status=SessionLog.START), step=0))
tensorboard.add_scalar('path/one/', 0, 0)
tensorboard.add_scalar('path/two/', 0, 0)
tensorboard.add_scalar('path/one/', val, 1)
tensorboard.add_scalar('path/two/', val, 1)
tensorboard.add_scalar('path/one/', 0, 2)
tensorboard.add_scalar('path/two/', 0, 2)
tensorboard.file_writer.event_writer.add_event(
    Event(session_log=SessionLog(status=SessionLog.STOP), step=2))
tensorboard.close()

sleep(1.0)

tensorboard = SummaryWriter(log_dir='welp/train')
tensorboard.file_writer.event_writer.add_event(Event(step=1, file_version='brain.Event:2'))
tensorboard.file_writer.event_writer.add_event(
    Event(session_log=SessionLog(status=SessionLog.START), step=1))
tensorboard.add_scalar('path/one/', 0, 1)
tensorboard.add_scalar('path/two/', 0, 1)
tensorboard.file_writer.event_writer.add_event(
    Event(session_log=SessionLog(status=SessionLog.STOP), step=1))
tensorboard.close()
```

**Comment by PetrochukM** [comment]
Is it possible for this to be run on initialization of tensorboard?

**Comment by lanpa** [comment]
You should not call low level api from outside. The result of the provided code is expected.
```
tensorboard = SummaryWriter(log_dir='welp/train')
tensorboard.file_writer.add_session_log(
    session_log=SessionLog(status=SessionLog.START), global_step=1)
tensorboard.add_scalar('path/one/', 0, 1) <-----
tensorboard.add_scalar('path/two/', 0, 1) <----- global_step should bigger than the last used global step (which is 3)
```

Since you are viewing the result of second code with `time passed` as horizontal axis. The above problem of wrongly use `global_step` will not be a observed.

**Comment by PetrochukM** [comment]
Umm, that’s not how TensorBoard works. The problem, i’m solving, overlapping events due to a system crash. TensorBoard allows you to restart at earlier checkpoint with a lower global step than the previous global step.

TensorBoard then purges events within the overlap. Your API does not support this feature of TensorBoard because your API fails to register file_version correctly.

**Comment by lanpa** [comment]
OK. This is worth implementing.

**Comment by lanpa** [comment]
done. please see https://github.com/lanpa/tensorboard-pytorch/blob/master/examples/demo_purge.py for example. (need to install from source at this time point)
